scheduler/src/cook/compute_cluster.clj

```diff
@@ -0,0 +1,124 @@
+  "Process one mesos cluster specification, returning the entity id of the corresponding compute-cluster,
+  creating the cluster if it does not exist."
+  [conn {:keys [compute-cluster-name framework-id] :as mesos-cluster}]
+  {:pre [compute-cluster-name
+         framework-id]}
+  (let [cluster-entity (get-mesos-cluster-entity-id (d/db conn) mesos-cluster)]
+    (when-not cluster-entity
+      (write-compute-cluster conn (mesos-cluster->compute-cluster-map-for-datomic mesos-cluster)))
+    {:cluster-type :mesos-cluster
+     :compute-cluster-name compute-cluster-name
+     :mesos-framework-id framework-id
+     :db-id (get-mesos-cluster-entity-id (d/db conn) mesos-cluster)}))
+
+(defn get-mesos-clusters-from-config
+  "Get all of the mesos clusters defined in the configuration.
+  In config.edn, we put all of the mesos keys under one toplevel dictionary.
+
+  E.g.:
+
+  {:failover-timeout-ms nil
+   :framework-id #config/env \"COOK_FRAMEWORK_ID\"
+   :master #config/env \"MESOS_MASTER\"
+   ...
+   }
+
+  However, in config.clj, we split this up into lots of different keys at the toplevel:
+
+  :mesos-master (fnk [[:config {mesos nil}]]
+      ...)
+  :mesos-framework-id (fnk [[:config {mesos ....
+
+  This function undoes this shattering of the :mesos {...} into separate keys that
+  occurs in config.clj. Long term, we need to fix config.clj to not to that, probably
+  as part of global cook, at which time, this probably won't need to exist. Until then however....."
+  [{:keys [mesos-compute-cluster-name mesos-framework-id]}]
+  [{:compute-cluster-name mesos-compute-cluster-name :framework-id mesos-framework-id}])
+
+(def cluster-name->cluster-dict-atom (atom nil))
+
+(defn cluster-name->db-id
+  "Given a cluster name, return the db-id we should use to refer to that compute cluster
+  when we put it within a task structure."
+  [cluster-name]
+  (let [{:keys [db-id]} (get @cluster-name->cluster-dict-atom cluster-name)]
+    ; All clusters referenced by name must have been installed in the db previously.
+    (when-not db-id (throw (IllegalStateException. (str "Was asked to lookup db-id for " cluster-name " and got nil"))))
+    db-id))
+
+(defn setup-cluster-map-config
+  "Setup the cluster-map configs, linking a cluster name to the associated metadata needed
+  to represent/process it."
+  [conn settings]
+  (let [compute-clusters (->> (get-mesos-clusters-from-config settings)
+                              (map (partial get-mesos-cluster-map conn)))
+        reduce-fn (fn [accum {:keys [compute-cluster-name] :as cluster-dict}]
+                    (when (contains? accum compute-cluster-name)
+                      (throw (IllegalArgumentException.
+                               (str "Multiple compute-clusters have the same name: " compute-cluster-name))))
+                    (assoc accum compute-cluster-name cluster-dict))]
+    (doseq [compute-cluster compute-clusters]
+      (log/info "Setting up compute cluster: " compute-cluster))
```

Comment:
Why not:

```clojure
(run! (fn [compute-cluster] (log/info "Setting up compute cluster: " compute-cluster)) compute-clusters)
```

Reply:
I can do that, but I'm curious to know why you're judging it as better? Its 10 bytes longer, its got more parenthesis and requires defining a new function.

Reply:
Seems like `run!` is precisely made for this sort of circumstance, whereas `doseq` is more general.

https://clojuredocs.org/clojure.core/run%21